File: backend-java/minimarket-api-java/src/main/java/com/minimarket/api/repository/PurchaseRepository.java
package com.minimarket.api.repository;

import com.minimarket.api.entity.Purchase;
import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.List;
import java.util.Optional;

public interface PurchaseRepository extends JpaRepository<Purchase, Integer> {

    @EntityGraph(attributePaths = {"supplier", "user", "details", "details.product"})
    List<Purchase> findAllByOrderByPurchaseDateDesc();

    @EntityGraph(attributePaths = {"supplier", "user", "details", "details.product"})
    Optional<Purchase> findById(Integer id);
}

File: backend/Minimarket.Api/Repositories/PurchaseRepository.cs
using Microsoft.EntityFrameworkCore;
using Minimarket.Api.Data;
using Minimarket.Api.Models;

namespace Minimarket.Api.Repositories;

public class PurchaseRepository(MinimarketDbContext context) : IPurchaseRepository
{
    public Task<List<Purchase>> GetAllAsync() =>
        context.Purchases
            .Include(x => x.Supplier)
            .Include(x => x.User)
            .Include(x => x.Details)
                .ThenInclude(x => x.Product)
            .OrderByDescending(x => x.PurchaseDate)
            .ToListAsync();

    public Task<Purchase?> GetByIdAsync(int id) =>
        context.Purchases
            .Include(x => x.Supplier)
            .Include(x => x.User)
            .Include(x => x.Details)
                .ThenInclude(x => x.Product)
            .FirstOrDefaultAsync(x => x.Id == id);

    public Task AddAsync(Purchase purchase) => context.Purchases.AddAsync(purchase).AsTask();

    public Task<int> SaveChangesAsync() => context.SaveChangesAsync();
}
